client/src/api/block_builder/input_selection/core/requirement/amount.rs

```diff
@@ -73,41 +73,109 @@ pub(crate) fn amount_sums(
+    native_tokens_remainder: bool,
+    timestamp: u32,
+}
+
+impl AmountSelection {
+    fn new(input_selection: &InputSelection) -> Result<Self> {
+        let (inputs_sum, outputs_sum, inputs_sdr, outputs_sdr) = amount_sums(
+            &input_selection.selected_inputs,
+            &input_selection.outputs,
+            input_selection.timestamp,
+        );
+        let (remainder_amount, native_tokens_remainder) = input_selection.remainder_amount()?;
+
+        Ok(Self {
+            newly_selected_ids: HashSet::new(),
+            newly_selected_inputs: Vec::new(),
+            inputs_sum,
+            outputs_sum,
+            inputs_sdr,
+            outputs_sdr,
+            remainder_amount,
+            native_tokens_remainder,
+            timestamp: input_selection.timestamp,
+        })
+    }
 
-        if remainder_amount > diff {
-            remainder_amount - diff
+    fn missing_amount(&self) -> u64 {
+        // If there is already a remainder, make sure it's enough to cover the storage deposit.
+        if self.inputs_sum > self.outputs_sum {
+            let diff = self.inputs_sum - self.outputs_sum;
+
+            if self.remainder_amount > diff {
+                self.remainder_amount - diff
+            } else {
+                0
+            }
+        } else if self.inputs_sum < self.outputs_sum {
+            self.outputs_sum - self.inputs_sum
+        } else if self.native_tokens_remainder {
+            self.remainder_amount
         } else {
             0
         }
-    } else if inputs_sum < outputs_sum {
-        outputs_sum - inputs_sum
-    } else if native_tokens_remainder {
-        remainder_amount
-    } else {
-        0
+    }
+
+    fn fulfil<'a>(&mut self, inputs: impl Iterator<Item = &'a InputSigningData>) -> bool {
+        for input in inputs {
+            if self.newly_selected_ids.contains(input.output_id()) {
+                continue;
+            }
+
+            if let Some(sdruc) = sdruc_not_expired(&input.output, self.timestamp) {
+                // Skip if no additional amount is made available
+                if input.output.amount() == sdruc.amount() {
+                    continue;
+                }
+                let input_sdr = self.inputs_sdr.get(sdruc.return_address()).unwrap_or(&0) + sdruc.amount();
+                let output_sdr = *self.outputs_sdr.get(sdruc.return_address()).unwrap_or(&0);
+
+                if input_sdr > output_sdr {
+                    let diff = input_sdr - output_sdr;
+                    self.outputs_sum += diff;
+                    *self.outputs_sdr.entry(*sdruc.return_address()).or_default() += sdruc.amount();
+                }
+
+                *self.inputs_sdr.entry(*sdruc.return_address()).or_default() += sdruc.amount();
+            }
+
+            self.inputs_sum += input.output.amount();
+            self.newly_selected_inputs.push((input.clone(), None));
+            self.newly_selected_ids.insert(*input.output_id());
```

Comment:
Is this better than HashMap<OutputId, InputSigningData>?

Reply:
You mean combining the hashset and the vec ?

Reply:
Yes

Reply:
Would have to be `HashMap<OutputId, (InputSigningData, Option<AliasTransition>)>` then, but I'll try yeah

Reply:
Yeah sure, just looks like it would be less error prone to combine them

Reply:
Done